[Mono] Initial metadata update support

lambdageek opened the pull request:
First part of support for metadata updates.

Contributes to https://github.com/dotnet/runtime/issues/44806

The feature is off by default.  To enable, build with `/p:MonoMetadataUpdate=true` (e.g. `./build.sh --os browser /p:MonoMetadataUpdate=true`).  There are samples in `src/mono/netcore/sample/mbr` (see the README - the samples aren't completely standalone and need some external tooling to build) for console (only tested on Mac and Linux) and wasm.

There's a demo at https://lambdageek.dev/dl0/

lambdageek committed:
[metadata_update] Keep track of logical table sizes for deltas

Keep track of inserted/modified rows for each table in each delta.
This will help to use a simpler algorithm to locate effective table rows by
keeping track of the logical number of rows in the appended tables

lambdageek committed:
Initial metadata-update prototype

Co-Authored-By: Bernhard Urban-Forster <lewurm@gmail.com>

lambdageek committed:
Add metadata-update.{c,h} to CMakeLists.txt

lambdageek committed:
fix icall-def-netcore typo

lambdageek committed:
Add icall to corelib

lambdageek committed:
Add console and browser metadata update samples

Both samples depend on the roslynildiff tool which should be specified with a
RoslynILDiffFullPath property in the .csproj files for the projects.

lambdageek committed:
Don't make entrypoint public yet

lambdageek committed:
[runtime] ifdef out metadata updates if not enabled

Also move execution engine initialization into the main update function and use
a MonoError to signal failures (such as if interp inlining is not turned off)
instead of asserting at startup.

lambdageek committed:
Add README for mbr samples

lambdageek committed:
[build] Add initial runtime support for MonoMetadataUpdate property

In the runtime defines cmake ENABLE_METADATA_UPDATE option and sets a preprocessor flag.

In System.Private.CoreLib, defines FEATURE_METADATA_UPDATE and uses it to throw
a NotSupportedException from LoadMetadataUpdate

lambdageek committed:
[wasm] set log mask to metadata-update

lambdageek committed:
[mbr] Add InjectUpdate fn to sample

lambdageek committed:
[metadata-update] don't merge heaps

lambdageek committed:
Add LoadMetadataUpdate to linker descriptor

lambdageek committed:
[wasm] add default Makefile variable value

lambdageek committed:
fix mono/mono CI

don't try to run enc tests yet since they depend on roslynildiff

lambdageek committed:
remove mono/mono/tests/enc

Will add as runtime tests in a future PR

lambdageek committed:
[metadata-update] Add per-thread exposed generation

A thread has to voluntarily roll up to the latest published generation in order
to see updates.

- Roll up to the latest published generation when attaching a thread
- The updater thread sees the allocated unpublished generation

lambdageek committed:
[mbr] Fixup console sample

Use a single changing testfile

lambdageek committed:
[metadata-update] delete unused method

lambdageek committed:
[mbr] Use 2 threads in console sample

lambdageek committed:
[metadata-update] Respect exposed generation in MethdDef RVA lookups

lambdageek committed:
[interp] Expose latest metadata update before transforming methods

lambdageek committed:
[mbr] Update samples after rebase

Use the WasmApp.targets

lambdageek committed:
[metadata-update] Don't fail after the first unsupported edit

Log all the unsupported edits, then cancel the update

lambdageek committed:
[metadata-update] Use a GList for MonoImage:delta_image

We're going to need to walk backwards from the latest published delta

lambdageek committed:
[metadata-update] add effective table lookup debug output

vargaz commented on src/mono/mono/metadata/icall-def-netcore.h:
```diff
@@ -326,6 +326,11 @@ HANDLES_REUSE_WRAPPER(MPROP_3, "get_metadata_token", ves_icall_reflection_get_to
+#ifdef ENABLE_METADATA_UPDATE
+ICALL_TYPE(RUNF, "System.Runtime.CompilerServices.RuntimeFeature", RUNF_1)
+HANDLES(RUNF_1, "LoadMetadataUpdate_internal", ves_icall_Mono_Runtime_LoadMetadataUpdate, void, 5, (MonoReflectionAssembly, gconstpointer, gint32, gconstpointer, gint32))
+#endif
```
Why does this need to be retained ?

vargaz commented on src/mono/mono/metadata/icall-def.h:
```diff
@@ -152,6 +152,9 @@ HANDLES(RUNTIME_2, "EnableMicrosoftTelemetry_internal", ves_icall_Mono_Runtime_E
 HANDLES(RUNTIME_12, "GetNativeStackTrace", ves_icall_Mono_Runtime_GetNativeStackTrace, MonoString, 1, (MonoException))
+#ifdef ENABLE_METADATA_UPDATE
+HANDLES(RUNTIME_22, "LoadMetadataUpdate_internal", ves_icall_Mono_Runtime_LoadMetadataUpdate, void, 5, (MonoReflectionAssembly, gconstpointer, gint32, gconstpointer, gint32))
+#endif
```
Not need on !netcore.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -6777,6 +6777,24 @@ ves_icall_Mono_Runtime_DumpStateTotal (guint64 *portable_hash, guint64 *unportab
+#ifdef ENABLE_METADATA_UPDATE
+void
+ves_icall_Mono_Runtime_LoadMetadataUpdate (MonoReflectionAssemblyHandle refassm,
+					   gconstpointer dmeta_bytes, int32_t dmeta_len,
```
Might be better to pass in the MonoAssembly* pointer.

vargaz commented on src/mono/mono/metadata/metadata-internals.h:
```diff
@@ -943,6 +976,18 @@ mono_metadata_clean_generic_classes_for_image (MonoImage *image);
+#else
+gboolean
+mono_metadata_table_bounds_check (MonoImage *image, int table_index, int token_index);
+#endif
```
This and others might need to be inline since they are perf critical.

lambdageek replied:
Ok inlined the fast path for `mono_metadata_table_bounds_check` and for `mono_image_effective_table`

lambdageek committed:
Address review feedback

lambdageek commented:
@vargaz Addressed the feedback.

@BrzVlad any thoughts about the interp changes?

vargaz commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml:
```diff
@@ -551,6 +551,11 @@
+		<type fullname="System.Runtime.CompilerServices.RuntimeFeature">
+			<method name="LoadMetadataUpdate" />
+		</type>
+
```
Why does this need to be retained ?

lambdageek replied:
Because it will only be called from a debugger session.

vargaz commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -466,10 +466,13 @@ EMSCRIPTEN_KEEPALIVE void
 	monoeg_g_setenv ("MONO_LOG_LEVEL", "debug", 0);
-	monoeg_g_setenv ("MONO_LOG_MASK", "gc", 0);
+	monoeg_g_setenv ("MONO_LOG_MASK", "metadata-update", 0);
     // Setting this env var allows Diagnostic.Debug to write to stderr.  In a browser environment this
```
Debug change.

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -7553,13 +7555,66 @@ invalidate_transform (gpointer imethod_)
+		if (!info)
+			continue;
+
+		MonoLMF *lmf = info->jit_data->lmf;
```
This fails to detect the currently executing methods (methods that are executing since the last interp entry)

lambdageek replied:
@BrzVlad how do I get the ones at the top of the stack?

BrzVlad replied:
I think the easiest for now would be to save the current frame in the `ThreadContext*` when we have a suspend request while in `MINT_SAFEPOINT`.

lambdageek replied:
@BrzVlad done, take another look, please.

lambdageek committed:
[interp] Save top interp frame at MINT_SAFEPOINT to ThreadContext

Give metadata updates a peek at the interp frames since the LMF so that it can
copy the InterpMethods that are currently executing

This only works with hybrid and full coop suspend.  Preemptive suspend will
need another mechanism.

lambdageek committed:
[mbr] Extend console sample

Add a busy thread to demonstrate that interp frames since the last managed
frame are visible to the metadata update mechanism and the active method bodies are
copied before being invalidated.

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -362,6 +362,7 @@ clear_resume_state (ThreadContext *context)
 	g_assert (context->exc_gchandle);
 	mono_gchandle_free_internal (context->exc_gchandle);
 	context->exc_gchandle = 0;
+	context->safepoint_frame = NULL;
```
Why null it here ? It should always be null when returning from `mono_threads_safepoint`

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -5218,9 +5235,11 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
 		MINT_IN_CASE(MINT_SAFEPOINT)
 			/* Do synchronous checking of abort requests */
 			EXCEPTION_CHECKPOINT;
+			context_set_safepoint_frame (context, frame);
```
MINT_SAFEPOINT is one the most common opcodes, we should try not to increase the work done on the fastpath. Could we check for mono_polling_required here ?

lambdageek committed:
[interp] Check mono_polling_required at safepoint

lambdageek commented:
@BrzVlad @vargaz anything else you'd like to see as part of this PR? Or anything else you think is important to follow-up on? If not, could one of your approve?
